clients/linux/src/input/key.rs

```diff
@@ -1,158 +1,188 @@
+        let state = x11::state_new_from_device(&keymap, conn, device_id);
 
-pub fn handle(
-    detail: u8, state: KeyButMask, pressed: bool, app: &mut WgpuLockbook,
-    paste_context: &mut input::clipboard_paste::Context,
-) -> Result<(), Box<dyn std::error::Error>> {
-    // "X11-style keycodes are offset by 8 from the keycodes the Linux kernel uses."
-    // https://github.com/rust-windowing/winit/blob/master/src/platform_impl/linux/common/keymap.rs#L7
-    let key = detail.saturating_sub(8);
+        Self { state }
+    }
 
-    let modifiers = modifiers(state);
+    pub fn handle(
+        &mut self, detail: u8, _state: KeyButMask, pressed: bool, app: &mut WgpuLockbook,
+        paste_context: &mut input::clipboard_paste::Context,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        let keycode = detail.into();
+        let keysym = self.state.key_get_one_sym(keycode);
+        let utf8 = xkb::keysym_to_utf8(keysym);
 
-    // text
-    if pressed && (modifiers.shift_only() || modifiers.is_none()) {
-        if let Some(text) = key_text(key, modifiers.shift) {
-            app.raw_input
-                .events
-                .push(egui::Event::Text(text.to_owned()));
+        if pressed {
+            self.state.update_key(keycode, xkb::KeyDirection::Down);
+        } else {
+            self.state.update_key(keycode, xkb::KeyDirection::Up);
         }
-    }
 
-    // todo: something feels weird about this
-    if let Some(key) = egui_key(key) {
-        if pressed && key == egui::Key::X && modifiers.command {
-            app.raw_input.events.push(egui::Event::Cut);
-        } else if pressed && key == egui::Key::C && modifiers.command {
-            app.raw_input.events.push(egui::Event::Copy);
-        } else if pressed && key == egui::Key::V && modifiers.command {
-            paste_context.handle_paste()?;
-        } else {
-            // other egui keys
-            app.raw_input.events.push(egui::Event::Key {
-                key,
-                pressed,
-                repeat: false,
-                modifiers,
-                physical_key: None,
-            });
+        let modifiers = Modifiers {
+            alt: self
+                .state
+                .mod_name_is_active(xkb::MOD_NAME_ALT, xkb::STATE_MODS_EFFECTIVE),
+            ctrl: self
+                .state
+                .mod_name_is_active(xkb::MOD_NAME_CTRL, xkb::STATE_MODS_EFFECTIVE),
+            shift: self
+                .state
+                .mod_name_is_active(xkb::MOD_NAME_SHIFT, xkb::STATE_MODS_EFFECTIVE),
+            command: self
+                .state
+                .mod_name_is_active(xkb::MOD_NAME_CTRL, xkb::STATE_MODS_EFFECTIVE),
+            mac_cmd: false,
+        };
+
+        // todo: pressing a key then holding ctrl or alt while releasing may cause egui to think it is still pressed
+        if (modifiers.is_none() || modifiers.shift_only()) && utf8.chars().any(|c| !c.is_control())
```

Comment:
somewhat hacky but works well in my testing and cleaner than the alternative (checking against a hardcoded set of control chars)

Reply:
Hard for me to determine the correctness of this

easy to determine the correctness of "this is how winit does it"